### PR TITLE
Date as text input in datePicker view.

### LINKED ISSUE #1418: SDC Catalog app: Date picker - Keyboard doesn't open when clicking text field
**Describe the bug**
When clicking the text field on the DatePicker or DateTime picker, the keyboard doesn't open. 

**Component**
SDC gallery app

**To Reproduce**
1. Go to SDC catalog app
2. Click on Date picker
3. Click on text field (not the calendar icon)

**Expected behavior**
The expected behavior is that manual keyboard based entry should be possible. We show a date format and only allow numerical date entry.

**Screenshots**
Screenshot_20220524-145927-1

### PR DESCRIPTION
**Pull request by santosh-pingle**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1418 

**Description**

- Enable textInputTextField to accept date as input.
- Support locale date pattern.

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Feature

**Screenshots (if applicable)**

https://user-images.githubusercontent.com/86107848/179512410-91690e01-04b9-448a-9e70-baa09dd16d9d.mov

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: datacapture/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -26,6 +26,7 @@
     <string name="submit_questionnaire">Submit</string>
     <string name="save">Save</string>
     <string name="date">Date</string>
+    <string name="hint_date_format">dd/mm/yyyy</string>
```
Comment (jingtang10) [inline datacapture/src/main/res/values/strings.xml:29]: this needs to be removed

File: datacapture/src/test/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryTest.kt [hunk 2]
```diff
@@ -207,6 +207,25 @@ class QuestionnaireItemDatePickerViewHolderFactoryTest {
       .isFalse()
   }
 
+  @Test
+  fun textInputEditText_acceptsDateInput() {
+    setLocale(Locale.US)
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryTest.kt:212]: add tests for a different locale

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt [hunk 3]
```diff
@@ -153,3 +186,28 @@ internal val DateType.localDate
...
+    DateTimeFormatterBuilder.getLocalizedDateTimePattern(
+      FormatStyle.SHORT,
+      null,
+      IsoChronology.INSTANCE,
+      Locale.getDefault()
+    )
+  val start = localDatePattern.indexOf('y')
+  val end = start + localDatePattern.filter { it == 'y' }.count()
+  return DateTimeFormatter.ofPattern(localDatePattern.replaceRange(start until end, "yyyy"))
+    .withLocale(Locale.getDefault())
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt:204]: i feel this is too custom. is there no way to just use a format style that is yyyy

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt (hunk 3 above)
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt:204]: can you try to use the android icu lib?

**Review by jingtang10 — CHANGES_REQUESTED** [review]

**Comment by santosh-pingle** [comment]
Code changes as per review comments are in progress.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryTest.kt (hunk 2 above)
Comment (santosh-pingle) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryTest.kt]: done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt (hunk 3 above)
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt]: done.

File: datacapture/src/main/res/values/strings.xml (hunk 1 above)
Comment (santosh-pingle) [reply datacapture/src/main/res/values/strings.xml:29]: done

**Comment by santosh-pingle** [comment]
https://user-images.githubusercontent.com/86107848/182585687-f636c8e0-25c5-429f-ace2-b556c74ce1f3.mov

File: datacapture/src/main/res/values/strings.xml (hunk 1 above)
Comment (jingtang10) [reply datacapture/src/main/res/values/strings.xml:29]: missing a commit to be pushed?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt [hunk 4]
```diff
@@ -80,16 +99,25 @@ internal object QuestionnaireItemDatePickerViewHolderFactory :
...
-          textInputLayout.helperText = it
-        }
+        var parseJob: Job? = null
+        textWatcher =
+          textInputEditText.doAfterTextChanged { text ->
+            // allow user to enter the input, e.g year may be y, yy, yyyy
+            parseJob?.cancel()
+            parseJob =
+              CoroutineScope(Dispatchers.Main).launch {
+                delay(1000)
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt:117]: this code that runs on the main thread seems fragile to me - i don't think we need this. simply updating answer within this callback seems enough?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt [hunk 5]
```diff
@@ -80,16 +99,25 @@ internal object QuestionnaireItemDatePickerViewHolderFactory :
...
             ?.localDate
             ?.localizedString
         )
-        questionnaireItemViewItem.questionnaireItem.entryFormat?.let {
-          textInputLayout.helperText = it
-        }
+        var parseJob: Job? = null
+        textWatcher =
+          textInputEditText.doAfterTextChanged { text ->
+            // allow user to enter the input, e.g year may be y, yy, yyyy
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt:113]: surely this is decided by the android api and not us?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt [hunk 6]
```diff
@@ -119,6 +147,24 @@ internal object QuestionnaireItemDatePickerViewHolderFactory :
...
+          val date = parseDate(text, textInputEditText.context.applicationContext)
+          localDate = date.toInstant().atZone(ZoneId.systemDefault()).toLocalDate()
+        } catch (e: ParseException) {
+          questionnaireItemViewItem.clearAnswer()
+        }
+
+        localDate?.run {
+          questionnaireItemViewItem.setAnswer(
+            QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent().apply {
+              value = DateType(localDate!!.year, localDate!!.monthValue - 1, localDate!!.dayOfMonth)
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt:163]: i'm sure in our code we do this in multiple places. can we extract this to an extension function.

```
fun LocalDate.toDateType(): DateType {
  ...
}
```

Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt:163]: and i'm sure i've seen the reverse in our code as well, but that can be done another time.

Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt:163]: this whole thing can probably be in the `try` block and then you don't have to have the `localDate` variable which you have to check if it's null.?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt [hunk 7]
```diff
@@ -44,12 +56,19 @@ internal object QuestionnaireItemDatePickerViewHolderFactory :
       private lateinit var textInputLayout: TextInputLayout
       private lateinit var textInputEditText: TextInputEditText
       override lateinit var questionnaireItemViewItem: QuestionnaireItemViewItem
+      private var textWatcher: TextWatcher? = null
+      private val localePattern =
+        DateTimeFormatterBuilder.getLocalizedDateTimePattern(
+          FormatStyle.SHORT,
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt:62]: nit: just add a comment here to say that medium and long formatstyles use date names which isn't what we want. we want numerical date format.

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+  private lateinit var viewHolder: QuestionnaireItemViewHolder
+  @Before
+  fun setup() {
+    activityScenarioRule.getScenario().onActivity { activity -> parent = FrameLayout(activity) }
+    viewHolder = QuestionnaireItemDatePickerViewHolderFactory.create(parent)
+    setTestLayout(viewHolder.itemView)
+  }
+
+  @Test
+  fun textInputEditText_dateInput_localeUs() = runBlocking {
```
Comment (jingtang10) [inline datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt:58]: use the more readable kotlin style in reverse quote marks as test name.

Comment (jingtang10) [inline datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt:58]: i think there're some other stuff we want to test... e.g.

- if you type invalid date, the answer will be cleared
- if you open dialog, and input a date, the text will be updated
- if you open dialog, and cancel, the text and answer won't change

possibly not everything is testable in this class. but try to add some of these test cases.

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt (hunk 7 above)
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt:59]: done.

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt (hunk 8 above)
Comment (santosh-pingle) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt]: It accepts quote but does not allow space, it shows identifier are not allowed in android projects.

Comment (santosh-pingle) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt]: Added unit and android tests.

**Comment by santosh-pingle** [comment]
https://user-images.githubusercontent.com/86107848/184305589-3f98d05d-2044-47fe-a60a-4bd9560b155c.mov

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt (hunk 8 above)
Comment (jingtang10) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt]: ok i didn't know this. thanks for trying. for the record: https://stackoverflow.com/questions/51562138/androidjunit4-doesnt-accept-space-function-test-name

this does seem to be an issue.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt [hunk 9]
```diff
@@ -44,12 +51,21 @@ internal object QuestionnaireItemDatePickerViewHolderFactory :
       private lateinit var textInputLayout: TextInputLayout
       private lateinit var textInputEditText: TextInputEditText
       override lateinit var questionnaireItemViewItem: QuestionnaireItemViewItem
+      private var textWatcher: TextWatcher? = null
+      // medium and long format styles use date names which isn't what we want. we want numerical
+      // date format.
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt:56]: optional nit:

```suggestion
      // Medium and long format styles use alphabetical month names which are difficult for the user to input. Use short format style which is always numerical.
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryTest.kt [hunk 10]
```diff
@@ -133,6 +133,67 @@ class QuestionnaireItemDatePickerViewHolderFactoryTest {
       .isEqualTo("11/19/20")
   }
 
+  @Test
+  fun `parse date text input as per US locale`() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryTest.kt:137]: nit:

```suggestion
  fun `parse date text input in US locale`() {
```

and below

**Review by jingtang10 — APPROVED** [review]

**Comment by santosh-pingle** [comment]
@jingtang10 
checking  below crash, if i enter date with 5 digit year e.g 21/12/19843 then it throws fhir data format exception in QuestionnaireViewModel. Same date is parsed as valid date  by android icu.

ca.uhn.fhir.parser.DataFormatException: Invalid date/time format: "19843-12-21": Expected character '-' at index 4 but found 3

**Comment by santosh-pingle** [comment]
> @jingtang10 checking below crash, if i enter date with 5 digit year e.g 21/12/19843 then it throws fhir data format exception in QuestionnaireItemViewItem. Same date is parsed as valid date by android icu.
> 
> ca.uhn.fhir.parser.DataFormatException: Invalid date/time format: "19843-12-21": Expected character '-' at index 4 but found 3

Crash happens in fhir library when it performs deep copy operation on [QuestionnaireResponse.QuestionnaireResponseItemComponent] which has above mentioned date.

**Comment by santosh-pingle** [comment]
> 

it also happens in QuestionnaireResponse deep copy operation.

**Comment by santosh-pingle** [comment]
> > @jingtang10 checking below crash, if i enter date with 5 digit year e.g 21/12/19843 then it throws fhir data format exception in QuestionnaireItemViewItem. Same date is parsed as valid date by android icu.
> > ca.uhn.fhir.parser.DataFormatException: Invalid date/time format: "19843-12-21": Expected character '-' at index 4 but found 3
> 
> Crash happens in fhir library when it performs deep copy operation on [QuestionnaireResponse.QuestionnaireResponseItemComponent] which has above mentioned date.

it also happens in QuestionnaireResponse deep copy operation.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt [hunk 11]
```diff
@@ -123,6 +143,27 @@ internal object QuestionnaireItemDatePickerViewHolderFactory :
...
+          // QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent in org.hl7.fhir.r4.model
+          // e.g ca.uhn.fhir.parser.DataFormatException: Invalid date/time format: "19843-12-21":
+          // Expected character '-' at index 4 but found 3
+          if (localDate.year.length() <= 4) {
+            questionnaireItemViewItem.setAnswer(
+              QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent().apply {
+                value = localDate.dateType
+              }
+            )
+          }
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt:162]: is it not better to have this logic in `parseDate` and throw a `ParseException` if the date is unparsable (even if it's by our definition)

**Review by jingtang10 — APPROVED** [review]

**Pull request merged**